Revisit out_data_decide_throttle

blueyed opened the pull request:
Pulse every 0.2s only, displaying first dot always ([1]).

This makes `:!yes` look much better (less busy) - although that might
have been intentional?!.

1: https://github.com/neovim/neovim/pull/10854#issuecomment-524858426

blueyed committed:
Revisit out_data_decide_throttle

Pulse every 0.1s only.

This makes `!yes` look much better (less busy).

justinmk commented:
Always showing the first dot makes the pulse look buggy. 

> This makes `:!yes` look much better (less busy

`yes` generates GB of data per minute, so I don't see why it should look less busy.

blueyed commented:
> Always showing the first dot makes the pulse look buggy.

ok, kept 4 states.

> `yes` generates GB of data per minute, so I don't see why it should look less busy.

But having indicator max. 10 times per second should be enough still, no?

The best would be to display some rate / throughput somehow in this case.

I was a bit confused by it displaying the output every 3s anyway - but think this is good after all.
Given the almost non-visible pulse (due to its flickering) I've thought that would be buggy somehow.

justinmk commented:
> But having indicator max. 10 times per second should be enough still, no?

Seems reasonable :)

> The best would be to display some rate / throughput somehow in this case.

Seems unnecessary. It won't be accurate, so it's just noise... and we still need a pulse anyways, to indicate activity.

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -405,22 +405,23 @@ static bool out_data_decide_throttle(size_t size)
+  size_t tick = visit % 4;
+  pulse_msg[0] = (tick > 0) ? '.' : ' ';
+  pulse_msg[1] = (tick > 1) ? '.' : ' ';
+  pulse_msg[2] = (tick > 2) ? '.' : ' ';
```
better, thanks. I think the comment can be unchanged though.

blueyed replied:
ok.

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -405,22 +405,23 @@ static bool out_data_decide_throttle(size_t size)
     }
+    if (since < (visit * 0.1 * NS_1_SECOND)) {
+      return true;
+    }
```
`if (visit % 20 == 0)` was to avoid spamming `os_hrtime()`.

blueyed replied:
..but we need it for rate limitting to 10/s.

justinmk commented on src/nvim/os/shell.c:
```diff
@@ -390,7 +390,7 @@ static bool out_data_decide_throttle(size_t size)
   static size_t     received    = 0;  // Bytes observed since last throttle.
   static size_t     visit       = 0;  // "Pulse" count of the current throttle.
-  static char       pulse_msg[] = { ' ', ' ', ' ', '\0' };
+  static char       pulse_msg[] = { '.', ' ', ' ', '\0' };
```
is this to help the screen test?

blueyed replied:
initially, but should be reverted (with 4 states again).

blueyed committed:
shell-test: remove REP_NODELAY, less delay with REP

REP_NODELAY was added because REP delayed too much.  This changes REP to
only add a delay on every 100th line instead.

This helps to cover the additional pulse steps with
out_data_decide_throttle, which would have required to change
REP_NODELAY anyway.

blueyed commented:
@justinmk 
Does https://github.com/neovim/neovim/pull/10855/commits/bd30930398eae5583d8a9ad413f54a94c1e0e132 look OK for you?
I'm a bit inclined to not touch the existing `REP`, but keep a separate action for what we need here - but on the other hand it makes `REP` a bit faster while keeping its behavior similar to what you wanted there.

justinmk reviewed (approved):
Pulse looks MUCH nicer on this branch! And ctrl-c also seems to react more quickly?! Thank you 😂

blueyed commented:
@justinmk 
So the change to "shell-test REP" looks OK to you?

> And ctrl-c also seems to react more quickly?!

Will check it - nice side effect then.. :)

justinmk commented:
> So the change to "shell-test REP" looks OK to you?

Reviewing 38a3af5dffd072e465fd, REP_NODELAY was added for a test that must reach the end of  output. For any other case, I would guess blasting the pty as fast as possible is unnecessary/unrealistic, and risks "starving" other threads/processes on slow CI.

Removing REP_NODELAY is appealing... we can try this change and tweak the rate maybe (if tests get flaky)?

> Could not see a difference with `:!grep "" . -R` though - what have you used for testing?

    :!yes

blueyed commented:
REP_NODELAY was added in bf8502300 for this test, since REP was not fast enough.
My main concern is that it might cause some regression now for where REP is too fast now, but overall I agree that we should try it.

blueyed commented:
btw: I cannot see a difference with regard to ctrl-c.
